### PR TITLE
[Flink-30755][runtime] Make SinkV2 support speculative execution for batch jobs

### PR DESCRIPTION
**Pull request by ifndef-SleePy**
## What is the purpose of the change

*This pull request introduces a way to allow SinkV2 run with speculative execution*

## Brief change log

  - *Introduce SupportsConcurrentExecutionAttempts interface*
  - *Add supportsConcurrentExecutionAttempts property for Transformation, StreamGraph, JobGraph*
  - *Make SinkV2 support speculative execution through implementing SupportsConcurrentExecutionAttempts interface*
  - *Use supportsConcurrentExecutionAttempts for excluding sink of speculative execution*

## Verifying this change

This change added tests:
  - *Add new test case in StreamingJobGraphGeneratorTest*
  - *Add new test case in SpeculativeSchedulerITCase*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (JavaDocs)

### CODE REVIEW

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGeneratorTest.java [hunk 1]
```diff
@@ -1733,6 +1745,176 @@ public void testCoordinatedSerializationException() {
...
+        // map1 and map2 are chained
+        // map1 supports concurrent execution attempt however map2 does not
+        streamGraph.setSupportsConcurrentExecutionAttempts(map1Node.getId(), true);
+        streamGraph.setSupportsConcurrentExecutionAttempts(map2Node.getId(), false);
+        streamGraph.setSupportsConcurrentExecutionAttempts(sinkNode.getId(), false);
+
+        final JobGraph jobGraph = StreamingJobGraphGenerator.createJobGraph(streamGraph);
+        assertThat(jobGraph.getNumberOfVertices()).isEqualTo(3);
+        for (JobVertex jobVertex : jobGraph.getVertices()) {
+            if (jobVertex.getName().contains("source")) {
```
Comment (alexandru-tetelea) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGeneratorTest.java:1785]: nit: What do you think about extracting these strings to constants?

**Review by zhuzhurk — COMMENTED** [review]
Thanks for creating this PR! @ifndef-SleePy 
The change looks generally good to me, except for the commit to remove legacy codes. I think it is needed but should be applied later. 
There are also a few minor comments. Please take a look.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java [hunk 2]
```diff
@@ -1389,27 +1388,6 @@ && isChainable(upStreamVertex.getInEdges().get(0), streamGraph)) {
...
-            vertexOperators.add(entry.getKey());
-            if (chainedConfigs.containsKey(entry.getKey())) {
-                vertexOperators.addAll(chainedConfigs.get(entry.getKey()).keySet());
-            }
-
-            for (int nodeId : vertexOperators) {
-                if (streamGraph.getSourceIDs().contains(nodeId)) {
-                    jobVertex.markContainsSources();
-                }
-                if (streamGraph.getSinkIDs().contains(nodeId)
```
[the remark this answers is not in the document]
Comment (ifndef-SleePy) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java:1404]: Yes, you are right. I think there is another easier way to do this. We can disable speculative execution for legacy sinks in `LegacySinkTransformation`.
By the way, I'm not sure the e2e test failure of Azure is caused by this issue or not. I would disable speculative execution through `LegacySinkTransformation` to re-trigger the Azure test again.

**Review by zhuzhurk — APPROVED** [review]
Thanks for addressing all the comments. @ifndef-SleePy 
LGTM.

**Pull request closed without merging**